tests/GenerateDocumentationTest.php

```diff
@@ -409,9 +409,22 @@ public function generates_correct_url_params_from_resource_routes_and_field_bind
         $this->artisan('scribe:generate');
 
         $groupA = Yaml::parseFile('.scribe/endpoints/00.yaml');
-        $this->assertEquals('providers/{provider_slug}/users/{user_id}/addresses', $groupA['endpoints'][0]['uri']);
+        $this->assertEquals('providers/{provider_slug}/users/{user}/addresses', $groupA['endpoints'][0]['uri']);
```

Comment:
The problem is that we don't want this. This sued to be the case before, but a number of people pointed out that it was unintuitive. It makes sense from the Laravel perspective, since we know about route model binding, but the reader of the API docs wants to know the value that goes in there (the user ID), not just the model.

Reply:
I agree that using only the model name is not intuitive, but that is not the case when using type-hinted variables in the controller action.
In the test case you pointed out, there are no type-hinted variables in the controller action, and that is the reason why the URL parameter was `user`.
But when using type-hinted variables in the controller action, the URL parameter will be `user_id` (or `user_slug` if the developer customized the route key name to `slug`). I think this behavior satisfies both the API docs reader and the developer since he can customize the route key name.

Reply:
Ah, I don't agree there. 

> since he can customize the route key name

Yes, but will he? It would be annoying to ask a developer to go through their routes and add `:id` for every parameter they use. Especially as Laravel defaults to using the `id` as the route key, it makes no sense that we can't.

Reply:
From your original comment:

> The bug is that Scribe uses the id as the default URL parameter when no inline binding is applied, but the user might be using other column by overriding the getRouteKeyName() method in the model.

I think we should focus on this. See if the model has `getRouteKeyName()` defined or a custom key in the route, and use that. Otherwise, we stick to `id`. We may just need to move the order of operations around to achieve this.

Reply:
So you want to stick with the `id` even for varaibles that are not type-hinted? as the `id` will be used by default for the type-hinted variables because `getRouteKeyName()` is defined in `Illuminate\Database\Eloquent\Model` to return the `id`.

Reply:
When you say "variables that are not type-hinted", what fo you mean?

Reply:
I mean the method arguments of the action that have types (`User $user` for example).

Reply:
If you're talking about a regular URL parameter, it doesn't get bound to a model by Laravel, and we don't touch it either. For instance, `/things/{thing}` should be as-is, unless it's a resource route or there's a `Thing $thing` parameter defined.

Reply:
Ok, I see your point.

As I understand, the required behavior is:
For non-resource routes, keep the current behavior of Scribe.
For resource routes:
- If there is a custom key in the route, use it,
- Else, if the method argument in the action is type-hinted call `getRouteKeyName()` in the model.
- Otherwise, use the `id`.

I will give it a try. Do you prefer I make the changes in a new clean pull request?

Reply:
Could you please confirm that this is the intended behavior?

Reply:
Could you give examples? It's been a while since I wrote this code, and I don't remember exactly the scenarios. It's easier to talk in code examples. Just short examples of a route for each case you've mentioned will do.

Reply:
Sure. The current behavior:
1. If there is an inline binding in the route, use it.
2. Else, if it is a resource route parameter, use the id.
3. Otherwise, keep the same parameter name.

What I will do in my new pull request, is to add a new step before the second step to check if there is a type-hinted variable in the controller action with the same name as the route segment name. This is the same behavior as the laravel [implicit binding](https://laravel.com/docs/9.x/routing#implicit-binding):

> Laravel automatically resolves Eloquent models defined in routes or controller actions whose type-hinted variable names match a route segment name.
---

#### Case 1: If there is an inline binding in the route.
```
Route::apiResource('posts', PostController::class)->only('show')->parameters(['posts' => 'post:slug']);
Route::get('posts/{post:slug}/comments', function() {});
```
Current behavior: use inline binding. Current URIs: `posts/{slug}` and `posts/{post_slug}/comments`.
The new behavior is the same.

#### Case 2: No inline binding exists, but model binding exists.
Assume that `Post@getRouteKeyName` returns `slug`, and `PostController@show` uses model binding (`Post $post`).
```
Route::apiResource('posts', PostController::class)->only('show').
Route::get('posts/{post}/comments', function(Post $post) {});
```
Current behavior: use the `id` for resource parameters, and keep the same name for non-resource parameters.
Current URIs: `posts/{id}` and `posts/{post}/comments`.
New behavior: call `getRouteKeyName()`.
New URIs: `posts/{slug}` and `posts/{post_slug}/comments`.

#### Case 3: Resource route parameter. No inline binding exists. No model binding exists.
Assume that `PostController@show` does not use model binding.
```
Route::apiResource('posts', PostController::class)->only('show');
```
Current behavior: use the `id`. Current URI: `posts/{id}`.
The new behavior is the same.

#### Case 4: Non-Resource route. No inline binding exists. No model binding exists.
```
Route::get('posts/{post}/commets', function($postId) {});
````
Current behavior: keep the same parameter name. Current URI: `posts/{post}`.
The new behavior is the same.

Reply:
I have another question about the testing, Is there a problem if I added 3 fixtures to test the new behavior?

The fixtures I want to add:
- `TestPost` model that overrides `getRouteKeyName()` to return `slug`.
- `TestPostController` with one method that has a type-hinted variable (`Post $post`).
- `TestPostUserController` with one method that uses two type-hinted variables (`Post $post, User $user`). I will use it to test nested resource routes.

Reply:
Yes, that sounds good. We probably need to clean up the code to make it less confusing, perhaps with some inline documentation of the examples. Also, I think the responsibility is half-shared; IIRC, there's somewhere else in the code where the URI is "fixed" slightly, which may change the results from here.

Reply:
Yeah, you can add whatever you need.